keps/sig-apps/3998-job-success-completion-policy/README.md

```diff
@@ -0,0 +1,604 @@
+	// "Indexed".
+	// Value can be an absolute number (ex: 5) or a percentage of total indexes
+	// when the job can be declared as succeeded (ex: 50%).
+	// The absolute number is calculated from the percentage by rounding up.
+	// 
+	// +optional
+	Count intstr.IntOrString
+
+	// Specifies the action to be taken on the not finished (succeeded or failed) pods 
+	// when the job achieved the requirements.
+	// Possible values are:
+	//
+	// - Continue: indicates that all pods wouldn't be terminated.
+	// - Terminate: indicates that not finished pods would be terminated.
+	WhenTargetAchieved WhenTargetAchievedSuccessPolicy
+}
+
+// WhenTargetAchievedSuccessPolicy specifies the action to be taken on the pods
+// when the job achieved the requirements.
+// +enum
+type WhenTargetAchievedSuccessPolicy string
+
+const (
+	// All pods wouldn't be terminated when the job reached successPolicy.
+	ContinueWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "Continue"
+
+	// Not finished pods would be terminated when the job reached successPolicy.
+	TerminateWhenTargetAchievedSuccessPolicy WhenTargetAchievedSuccessPolicy = "Terminate"
+)
+```
+
+Moreover, we validate the following constraints for the `onSucceededPods`:
+- exactly one of the fields `indexes` and `count` is specified
+  for a requirement.
+- whether the specified indexes in the `indexes` and
+  the number of indexes in the `count` don't exceed the value of `completions`.
+- whether `Indexed` is specified in the `completionMode` field.
+- whether the size of `indexes` is under 64Ki.
+- exactly one of the enum `continue` or `terminate` is specified
+in the `whenTargetAchieved`.
+
+In addition, when the `successPolicy` isn't nil, we validate the `onSucceededPods` isn't nil.
+
+### Evaluation
+We evaluate the SuccessPolicy every in updating pod conditions.
+And then if the job meets the SuccessPolicy, we add `Completed` condition
+instead of `Failed` condition to `.status.conditions` even if some indexes fail.
+
+If a set of indexes are specified in the `indexes` field,
+we evaluate `.status.completedIndexes` to see if a set of indexes is there.
+
+If the number of indexes is specified in the `count`, 
+we evaluate `.status.succeeded` to see if the value is `count` or more.
+
+Also, the job controller do actions against running pods based on the value of `onSucceededPods.whenTargetAchieved`:
+
+- `continue`: Nothing. Running pods continues to run.
+- `terminate`: Not finished pods would be terminated by the job controller. Not finished pods means running or pending pods.
+
+#### The situations where successPolicy conflicts other terminating policies
+
+The successPolicy has potential conflicts with the [pod failure policy](https://kubernetes.io/docs/tasks/job/pod-failure-policy/)
+and backoffLimit in the following situations: 
+
+- when the job meets the successPolicy and some pod failure policies with the `FailJob` action.
```

Comment:
What would we do in case of job exceeding activeDeadlineSeconds? Currently we fail the job, but it is reported as a bug. It makes sense to clarify here: https://github.com/kubernetes/kubernetes/issues/117303. Maybe we leave the decision for graduation to Beta?

Reply:
Yes it is separate. I only mentioned it because the list of possible failure reasons was mentioned and acivedeadline is one of them. However, feel free to resolve and handle separately from the keps

Reply:
Thank you for confirmation.
After this KEP is merged, I'm going to work on https://github.com/kubernetes/kubernetes/issues/117303.